torch/lib/c10d/ProcessGroup.cpp

```diff
@@ -92,18 +101,21 @@ OpType ProcessGroup::Work::retrieveOpType() {
 ProcessGroup::Work::~Work() {}
 
 bool ProcessGroup::Work::isCompleted() {
-  std::lock_guard<std::mutex> lock(mutex_);
-  return completed_;
+  return future_->completed();
 }
 
 bool ProcessGroup::Work::isSuccess() const {
-  std::lock_guard<std::mutex> lock(mutex_);
-  return !exception_;
+  return !future_->hasError();
+}
+
+void ProcessGroup::Work::setError(std::exception_ptr eptr) {
+  if (future_->exception_ptr() != nullptr) {
+    future_->setError(eptr);
+  }
```

Comment:
This is technically unsafe, because in principle someone else could set the error in between the check (on line 112) and the set (line 113). A safer way would be to use `setErrorIfNeeded` (though perhaps that's a private method? don't remember).

I also think that we should perhaps just throw we attempt to set an error on an already-errored future, rather than silently ignore it. I suspect that very often this would happen due to a programming bug (since typically there's only one "producer" responsible for completing a future) hence it would be good to give as much explicit early signal as possible about such a bug.